include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,338 @@
+
+    Info << "NSGA2::Optimize() Generation: " << generation << std::endl;
+    // have P_t, generate G_t using P_t
+    Info << "NSGA2::Optimize() BinaryTournamentSelection" << std::endl;
+    BinaryTournamentSelection(population);
+
+    // evaluate objectives
+    Info << "NSGA2::Optimize() EvaluateObjectives" << std::endl;
+    calculatedObjectives.resize(population.size());
+    EvaluateObjectives(population, objectives, calculatedObjectives);
+
+    // perform fast non dominated sort on $$ P_t \cup G_t $$
+    Info << "NSGA2::Optimize() FastNonDominatedSort" << std::endl;
+    ranks.resize(population.size());
+    FastNonDominatedSort(fronts, ranks, calculatedObjectives);
+
+    // perform crowding distance assignment
+    Info << "NSGA2::Optimize() CrowdingDistanceAssignment" << std::endl;
+    crowdingDistance.resize(population.size());
+    for (size_t fNum = 0; fNum < fronts.size(); fNum++)
+    {
+      CrowdingDistanceAssignment(fronts[fNum], objectives, crowdingDistance);
+    }
+
+    // sort based on crowding distance
+    Info << "NSGA2::Optimize() Sort(Crowding Distance)" << std::endl;
+    std::sort(population.begin(),
+              population.end(),
+              [this, ranks, crowdingDistance, population](MatType candidateP,
+                                                          MatType candidateQ)
+              {
+                size_t idxP, idxQ;
+                for(int i = 0; i < population.size(); i++)
+                {
+                  if (arma::approx_equal(population[i], candidateP, "absdiff", epsilon))
+                  {
+                    idxP = i;
+                  }
+                  if (arma::approx_equal(population[i], candidateQ, "absdiff", epsilon))
+                  {
+                    idxQ = i;
+                  }
+                }
+
+                return CrowdingOperator(idxP,
+                                        idxQ,
+                                        ranks,
+                                        crowdingDistance);
+              }
+    );
+
+    // yeild new population P_{t+1}
+    Info << "NSGA2::Optimize() Get P(t+1)" << std::endl;
+    population.resize(populationSize);
+  }
+
+  std::vector<MatType> bestFront;
+
+  for(size_t f: fronts[0])
+  {
+    bestFront.push_back(population[f]);
+  }
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+  return bestFront;
+}
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType>
+inline void NSGA2::EvaluateObjectives(std::vector<MatType> population,
+                                      MultiobjectiveFunctionType objectives,
+                                      std::vector<arma::vec>& calculatedObjectives)
+{
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    calculatedObjectives[i] = objectives.Evaluate(population[i]);
+  }
+}
+
+template<typename MatType>
+inline void NSGA2::BinaryTournamentSelection(std::vector<MatType>& population)
```

Comment:
Looks like we don't modify the population, so let's use `const std::vector<MatType>& population`.

Reply:
We are adding child nodes to the parent population near the end of the function!

Reply:
Ahh, thanks for the clarification.